### PR TITLE
[SPARK-26620][PYTHON] Make `DataFrameReader.json()` and `csv()` in Python should accept DataFrame.

### PR DESCRIPTION
**Pull request by ueshin**
## What changes were proposed in this pull request?

Currently `DataFrameReader.json()` and `csv()` in Python doesn't accept DataFrame, but they should accept if the schema of the given `DataFrame` contains only one String column.

## How was this patch tested?

Added corresponding tests and existing tests.

### CODE REVIEW

**Comment by ueshin** [comment]
cc @HyukjinKwon

**Comment by ueshin** [comment]
Jenkins, retest this please.

File: python/pyspark/sql/readwriter.py [hunk 1]
```diff
@@ -289,8 +294,14 @@ def func(iterator):
...
             return self._df(self._jreader.json(jrdd))
+        elif isinstance(path, DataFrame):
+            if len(path.schema) != 1 or not isinstance(path.schema[0].dataType, StringType):
+                raise TypeError("The DataFrame can contain only one String column: %s"
+                                % path.schema.simpleString())
+            jdataset = path._jdf.__getattr__('as')(self._spark._jvm.Encoders.STRING())
+            return self._df(self._jreader.json(jdataset))
         else:
-            raise TypeError("path can be only string, list or RDD")
+            raise TypeError("path can be only string, list, RDD or DataFrame")
```
Comment (HyukjinKwon) [inline python/pyspark/sql/readwriter.py]: Eh .. @ueshin, do you actually mind if we see how the discussion would go in https://github.com/apache/spark/pull/23534 before getting this in? I was thinking both are quite related.

Comment (srowen) [reply python/pyspark/sql/readwriter.py]: I'd also say this is working as intended. on the scala side it makes sense to support `Dataset[String]` and an `RDD[String]` is the closest analog here in Pyspark. It's easy to select a column out of a DataFrame anyway.

**Comment by srowen** [comment]
Same comment as on the related PR; I'd argue this is consistent now, but there are larger changes we could consider that change that consistent behavior.

**Pull request closed without merging**